stdlib/socket/src/main/java/org/ballerinalang/stdlib/socket/tcp/SelectorManager.java

```diff
@@ -0,0 +1,238 @@
+        return SelectorManagerHolder.manager;
+    }
+
+    /**
+     * Add channel to register pending socket queue. Socket registration has to be happen in the same thread
+     * that selector loop execute.
+     *
+     * @param callback A {@link ChannelRegisterCallback} instance which contains the resources,
+     *                      packageInfo and A {@link SelectableChannel}.
+     */
+    public void registerChannel(ChannelRegisterCallback callback) {
+        registerPendingSockets.add(callback);
+        selector.wakeup();
+    }
+
+    /**
+     * Unregister the given client channel from the selector instance.
+     *
+     * @param channel {@link SocketChannel} that about to unregister.
+     */
+    public void unRegisterChannel(SocketChannel channel) {
+        final SelectionKey selectionKey = channel.keyFor(selector);
+        if (selectionKey != null) {
+            selectionKey.cancel();
+        }
+    }
+
+    /**
+     * Start the selector loop.
+     */
+    public void start() {
+        if (running) {
+            return;
+        }
+        executor.execute(this::execute);
+        running = true;
+    }
+
+    private void execute() {
+        while (execution) {
+            try {
+                registerChannels();
+                final int select = selector.select();
+                if (select == 0) {
+                    continue;
+                }
+                Iterator<SelectionKey> keyIterator = selector.selectedKeys().iterator();
+                while (keyIterator.hasNext()) {
+                    SelectionKey key = keyIterator.next();
+                    keyIterator.remove();
+                    performAction(key);
+                }
+            } catch (Throwable e) {
+                log.error("An error occurred in selector loop: " + e.getMessage(), e);
+            }
+        }
+    }
+
+    /*
+    Channel registration has to be done in the same thread that selector loops runs.
+     */
+    private void registerChannels() {
+        ChannelRegisterCallback channelRegisterCallback;
+        while ((channelRegisterCallback = registerPendingSockets.poll()) != null) {
+            try {
+                SocketService socketService = channelRegisterCallback.getSocketService();
+                socketService.getSocketChannel()
+                        .register(selector, channelRegisterCallback.getInitialInterest(), socketService);
+            } catch (ClosedChannelException e) {
+                channelRegisterCallback.notifyFailure("Socket already closed");
+            }
+            channelRegisterCallback
+                    .notifyRegister(channelRegisterCallback.getInitialInterest() == SelectionKey.OP_READ);
+        }
+    }
+
+    private void performAction(SelectionKey key) {
+        if (!key.isValid()) {
+            key.cancel();
+        } else if (key.isAcceptable()) {
+            onAccept(key);
+        } else if (key.isReadable()) {
+            onReadReady(key);
+        }
```

Comment:
Are these 3 the only possible states? Should there be any assertion error in case the state is something else?

Reply:
isConnectable and isWritable are the other available state. We don't need isConnectable for the moment and isWritable will implement after this initial release. There is no state for the error situation.

Reply:
My concern is, in case the state is something else, this method will silently execute without any indication of what happened.
If that is the expected behavior then no issue.

Reply:
Yes. Those states will not execute with the current implementation. But this will address with the next round of improvements.